pkg/pillar/cmd/zfsmanager/handlediskconfig.go

```diff
@@ -0,0 +1,136 @@
+		}
+	} else {
+		// we can receive partition number
+		devName, err := types.GetDiskNameByPartName(dev.Name)
+		if err != nil {
+			log.Errorf("failed to get disk name by part name %s: %s", dev.Name, err)
+		}
+		if strings.Contains(devName, name) {
+			return devName, dev.Stat
+		}
+	}
+	return "", libzfs.VDevStat{}
+}
+
+func handleDiskConfigRestart(ctxArg interface{}, restartCounter int) {
+	log.Functionf("handleDiskConfigRestart(%d)", restartCounter)
+	ctx := ctxArg.(*zfsContext)
+	select {
+	case ctx.disksProcessingTrigger <- struct{}{}:
+	default:
+		log.Functionf("handleDiskConfigRestart: disksProcessingTrigger already triggered")
+	}
+}
+
+func processDisksTask(ctx *zfsContext) {
+
+	disksProcessingTicker := time.NewTicker(disksProcessingInterval)
+
+	wdName := agentName + "diskstask"
+	stillRunning := time.NewTicker(stillRunningInterval)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+
+	for {
+		select {
+		case <-disksProcessingTicker.C:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-ctx.disksProcessingTrigger:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func processDisks(ctx *zfsContext) error {
+	if vault.ReadPersistType() != types.PersistZFS {
+		return nil
+	}
+	disksConfig := ctx.subDiskConfig.GetAll()
+	if len(disksConfig) == 0 {
+		return nil
+	}
+	persistPool, err := libzfs.PoolOpen(vault.DefaultZpool)
+	if err != nil {
+		return fmt.Errorf("cannot open pool: %s", err)
+	}
+	defer persistPool.Close()
+	vdevTree, err := persistPool.VDevTree()
+	if err != nil {
+		return fmt.Errorf("cannot get vdev tree: %s", err)
+	}
+	for _, cfg := range disksConfig {
+		diskCfg := cfg.(types.EdgeNodeDiskConfig)
+		diskName := filepath.Base(diskCfg.Disk.Name)
+		if diskCfg.OldDisk != nil {
+			oldDiskName := filepath.Base(diskCfg.OldDisk.Name)
+			oldDevName, oldDevStat := getVdevNameStatus(vdevTree, oldDiskName)
+			if oldDevName != "" && diskCfg.Disk.Name != "" {
+				// if we found device oldDevName, replace it
+				log.Functionf("replacing %s with %s, old stat %s", oldDevName, diskCfg.Disk.Name, oldDevStat.State.String())
+				if stdout, err := zfs.ReplaceVDev(log, vault.DefaultZpool, oldDevName, diskCfg.Disk.Name); err != nil {
+					log.Errorf("cannot replace %s with %s: %s %s", oldDevName, diskCfg.Disk.Name, stdout, err)
+					continue
+				}
+			}
+		}
+		devName, devStat := getVdevNameStatus(vdevTree, diskName)
+		// found in pool
+		if devName != "" {
+			log.Functionf("zpool config disk %s, op %d, stat %s", devName, diskCfg.Config, devStat.State.String())
+			switch diskCfg.Config {
+			case types.EdgeNodeDiskConfigTypeUnspecified:
+				continue
+			case types.EdgeNodeDiskConfigTypeZfsOnline:
+				if zfs.GetZfsDeviceStatusFromStr(devStat.State.String()) == info.StorageStatus_STORAGE_STATUS_ONLINE {
+					continue
```

Comment:
Shouldn't this be an error? E.g., if trying to online a device which is unused.

Reply:
Thus a case on the status would make more sense where already online is a no-op, and offline is ok, and other status is an error.

Reply:
Fixed